### PR TITLE
[DEV-3674] Add support for BigQuery

### PR DESCRIPTION
**Pull request by kchua78**
## Description

Add support for BigQuery

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/session/base.py [hunk 1]
```diff
@@ -1265,9 +1284,7 @@ async def update_metadata_schema_version(self, new_version: int) -> None:
         new_version : int
             New version to update the working schema to
         """
-        update_version_query = (
-            f"""UPDATE METADATA_SCHEMA SET WORKING_SCHEMA_VERSION = {new_version}"""
-        )
+        update_version_query = f"""UPDATE {self.session.metadata_schema} SET WORKING_SCHEMA_VERSION = {new_version} WHERE 1=1"""
```
Comment (YS-L) [inline featurebyte/session/base.py:1287]: What is the reason to add the `1=1` condition?

File: featurebyte/session/bigquery.py [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+                    metadata={ARROW_METADATA_DB_VAR_TYPE: db_var_type},
+                )
+            )
+        return pa.schema(fields)
+
+    def fetch_query_result_impl(self, cursor: Any) -> pd.DataFrame | None:
+        """
+        Fetch the result of executed SQL query from connection cursor
+
+        This is an implementation specific to Snowflake that is more efficient when applicable.
```
Comment (YS-L) [inline featurebyte/session/bigquery.py:296]: Should the docstring be updated?

File: featurebyte/session/bigquery.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+        return column_name_type_map
+
+    async def get_table_details(
+        self,
+        table_name: str,
+        database_name: str | None = None,
+        schema_name: str | None = None,
+    ) -> TableDetails:
+        query = (
+            f'SELECT * FROM "{database_name}"."INFORMATION_SCHEMA"."TABLES" WHERE '
```
Comment (YS-L) [inline featurebyte/session/bigquery.py:377]: Would it be better to only select the specific columns that we need in this query?

File: featurebyte/session/bigquery.py [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+        })
+        return TableDetails(
+            details=json.loads(details.iloc[0].to_json(orient="index")),
+            fully_qualified_name=sql_to_string(
+                fully_qualified_table_name, source_type=self.source_type
+            ),
+        )
+
+    @staticmethod
+    def get_columns_schema_from_dataframe(  # pylint: disable=too-many-branches
```
Comment (YS-L) [inline featurebyte/session/bigquery.py:397]: Is this exactly the same as the snowflake's implementation?

File: featurebyte/common/documentation/documentation_layout.py [hunk 5]
```diff
@@ -852,6 +852,7 @@ def _get_utility_classes_layout() -> List[DocLayoutItem]:
             doc_path_override="api.groupby.GroupBy.forward_aggregate_asat.md",
         ),
         DocLayoutItem([UTILITY_CLASSES, FEATURE, "FeatureVersionInfo"]),
+        DocLayoutItem([UTILITY_CLASSES, WAREHOUSE, "BigQueryDetails"]),
```
Comment (YS-L) [inline featurebyte/common/documentation/documentation_layout.py:855]: If we don't expose this in the documentation and don't expose the new class in the top level init file, would that allow us to merge changes to main branch incrementally? I think that can simplify the development process.

File: featurebyte/session/base.py (hunk 1 above)
Comment (kchua78) [reply featurebyte/session/base.py:1287]: BigQuery requires a condition for UPDATE statements

File: featurebyte/session/bigquery.py (hunk 3 above)
Comment (kchua78) [reply featurebyte/session/bigquery.py:377]: This is actually from the snowflake implementation (and the functions below). Will work on the actual implementation and add test coverage

File: featurebyte/session/bigquery.py [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+        cursor: Any
+            Cursor to fetch data from
+
+        Returns
+        -------
+        pa.Schema
+        """
+        # get schema from query job as cursor description is missing REPEATED information
+        try:
+            schema = self._client._get_query_results(cursor.query_job.job_id, DEFAULT_RETRY).schema  # pylint: disable=protected-access
```
Comment (YS-L) [inline featurebyte/session/bigquery.py:310]: Does `_get_query_results` retrieve rows as well? Will it be an issue if the query result is large?

File: featurebyte/session/bigquery.py [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+
+        table_ref = TableReference(
+            dataset_ref=DatasetReference(project=self.project_name, dataset_id=self.dataset_name),
+            table_id=table_name,
+        )
+        job_config = LoadJobConfig(
+            schema=table_schema,
+            write_disposition="WRITE_EMPTY",
+        )
+        job = self._client.load_table_from_json(
```
Comment (YS-L) [inline featurebyte/session/bigquery.py:493]: Would be good to document the reason of using json as the format here.

**Review by YS-L — APPROVED** [review]

File: featurebyte/session/bigquery.py (hunk 6 above)
Comment (kchua78) [reply featurebyte/session/bigquery.py:310]: Indeed I will set `page_size` to 1

**Pull request merged**